### ISSUE TITLE
rendering_threads parameter not exported in the appleseed project file

### ISSUE THREAD

**Issue by Mango-3** [appleseedhq/blenderseed#232]
Setting/limiting the render threads manually in Blender works well for internal rendering but is not exported in the appleseed project file. After loading an exported file into Appleseed Studio the project defaults to the Auto setting using all available cpu cores.

**Comment by jdent02** [comment]
Thanks Mange.  I'll get it added in.

**Comment by jdent02** [comment]
#231

**Comment by Mango-3** [comment]
Also this one works now :))

**Comment by est77** [comment]
I removed the export of the number of threads to use in projects from Maya some time ago after we opened this ticket: https://github.com/appleseedhq/appleseed-maya/issues/75

I think we should be consistent in all plugins.

**Comment by Mango-3** [comment]
I don't see #75 as a problem. With no intervention the default is to use all logical cores but if I purposely set the thread number lower I would like to have it written into the project file. As a non-professional I don't have a dedicated render workstation or render farm. Limiting the render threads allows me to continue using my laptop while a render image is calculated.

**Comment by Mango-3** [comment]
Sure Esteban, I see your point. It should be consistent in all plugins. If you give away the project file you have some more things to consider. The recipient probably needs to use the same version of Appleseed as the project file was written, otherwise inconsistencies may arise. For instance until recently there was an as_layer shader in Appleseed, now not anymore. I haven't tried it but probably a project file referencing the as_layer shader won't render with the latest version of Appleseed.

**Comment by dictoon** [comment]
@Mango-3 The solution in your case is to set the default number of threads in the settings file of appleseed.cli or appleseed.studio (found in the `settings/` folder of your appleseed installation):
```xml
<parameter name="rendering_threads" value="auto" />
```

Regarding OSL shaders: you're right in the sense that you almost always want to use the exact same version on all your render machines to ensure consistency but that's really a different issue than the one being discussed here: not all your machines are going to have the same number of CPU cores so it makes sense to keep this setting local to each machine instead of forcing it in project files.

What do you think?

**Comment by Mango-3** [comment]
I agree François, setting the render threads in the settings/ folder is overall the better solution.
I just wanted to point out that when sharing a project file one cannot be certain that the recipient can just press the render button without checking the scene and settings and everything is already properly set.

**Comment by dictoon** [comment]
> I agree François, setting the render threads in the settings/ folder is overall the better solution.

Ok, we'll revert this change then.

> I just wanted to point out that when sharing a project file one cannot be certain that the recipient can just press the render button without checking the scene and settings and everything is already properly set.

We're clearly striving for that: It must be easy to share scenes and allow others to render them without fiddling with anything beyond maybe search paths (which, like render threads, should be configured locally on the individual machines).

In fact, until recently it was clearly the case. The introduction of OSL shaders complicated this somewhat, in particular the fact that our OSL shader library is very much work-in-progress.

My plan for the near future (which has yet to be discussed with the team) is to simply freeze whatever OSL shaders we have at this point, and then introduce changes or new shaders alongside the existing, frozen ones.

At some later point legacy OSL shaders will be deprecated and migration to the newer shaders will be encouraged. At some even later point legacy OSL shaders will be removed but users will always have the possibility to store a copy of the shaders alongside their scenes if they haven't migrated.

**Comment by Mango-3** [comment]
@dictoon 
Though it's not directly related to this ticket I share my thoughts on your last comments.
It's clearly a tricky issue to keep compatibility with project files. I see two ways to approach it.
1) The one you outline, keeping several generations of shader with similar functions. If its well documented not a problem. When I used Max I wondered why there are e.g. two versions of Disney shader. It was not immediately clear. Later I found out that only the older ones can be used with the Max procedural shader and only after setting a compatibility flag. 
2)  Have a project converter program which converts older versions of project files to the current one. Such a program may be called automatically upon loading an older project file into Studio or cli. Then only one set of shader would be required provided that the new ones are a super-set of the older.

**Comment by dictoon** [comment]
I completely agree, we need to document things properly as to avoid any confusion.

The compatibility flag ("Use Max Procedural Maps") may be removed in a future version of the plugin, while retaining the possibility to (automatically) use Max procedural maps with built-in materials (and only those; procedural maps aren't compatible with OSL shaders).

We have a project converter already, built into appleseed (you'll see migration messages in appleseed.studio's log panel if you load older projects) and in the form of a standalone command line tool (`projecttool` in `bin/`). Automatic migrations do not take care of migrating from one version of an OSL shader to a newer one yet because in the general case, it isn't possible to do it automatically (but it could be done for simple cases).

**Comment by Mango-3** [comment]
@dictoon Thanks for the detailed explanation.

**Comment by jdent02** [comment]
Should we consider this closed then?

**Comment by Mango-3** [comment]
@jdent02 Yep, from my side all is good. Sorry you had 2 x the work but I think the discussion was helpful for all involved.
